pilot/pkg/networking/core/v1alpha3/gateway.go

```diff
@@ -403,90 +397,157 @@ func buildGatewayListenerTLSContext(server *networking.Server) *auth.DownstreamT
-// them by this server's port and hostnames, and produces network filters for each destination from the filtered services
-func buildGatewayNetworkFilters(node *model.Proxy, env *model.Environment, server *networking.Server, gatewayNames map[string]bool) []listener.Filter {
+// buildGatewayNetworkFiltersFromTLSRoutes builds tcp proxy routes for all VirtualServices with TCP blocks.
+// It first obtains all virtual services bound to the set of Gateways for this workload, filters them by this
+// server's port and hostnames, and produces network filters for each destination from the filtered services
+func buildGatewayNetworkFiltersFromTCPRoutes(node *model.Proxy, env *model.Environment, server *networking.Server,
+	gatewaysForWorkload map[string]bool) []listener.Filter {
 	port := &model.Port{
 		Name:     server.Port.Name,
 		Port:     int(server.Port.Number),
 		Protocol: model.ParseProtocol(server.Port.Protocol),
 	}
 
-	dests := getVirtualServiceTCPDestinations(env, server, gatewayNames)
-	// de-dupe destinations by hostname; we'll take a random destination if multiple claim the same host
-	byHost := make(map[model.Hostname]*networking.Destination, len(dests))
-	for _, dest := range dests {
-		byHost[model.Hostname(dest.Host)] = dest
+	gatewayServerHosts := make(map[model.Hostname]bool, len(server.Hosts))
+	for _, host := range server.Hosts {
+		gatewayServerHosts[model.Hostname(host)] = true
 	}
 
-	filters := make([]listener.Filter, 0, len(byHost))
-	for host, dest := range byHost {
-		upstream, err := env.GetService(host)
-		if err != nil || upstream == nil {
-			log.Debugf("failed to retrieve service for destination %q: %v", host, err)
+	virtualServices := env.VirtualServices(gatewaysForWorkload)
+	var upstream *networking.Destination
+	for _, spec := range virtualServices {
+		vsvc := spec.Spec.(*networking.VirtualService)
+		matchingHosts := pickMatchingGatewayHosts(gatewayServerHosts, vsvc.Hosts)
+		if len(matchingHosts) == 0 {
+			// the VirtualService's hosts don't include hosts advertised by server
 			continue
 		}
-		filters = append(filters, buildOutboundNetworkFilters(node,
-			istio_route.GetDestinationCluster(dest, upstream, int(server.Port.Number)), "", port)...)
+
+		// ensure we satisfy the rule's l4 match conditions, if any exist
+		// For the moment, there can be only one match that succeeds
+		// based on the match port/server port and the gateway name
+		for _, tcp := range vsvc.Tcp {
+			if l4MultiMatch(tcp.Match, server, gatewaysForWorkload) {
+				upstream = tcp.Route[0].Destination // We pick first destination because TCP has no weighted cluster
+				break
+			}
+		}
+
+		// for opaque TCP services, the first TCP match wins as we can setup
+		// a listener on a port and forward to a single backend cluster
+		if upstream != nil {
+			break
+		}
 	}
-	return filters
+
+	destSvc, err := env.GetService(model.Hostname(upstream.Host))
```

Comment:
upstream can be nil here and this will panic. Add a guard and return if upstream == nil after the loop